File: pyps/src/main/scala/com/tencent/angel/ps/tensor/TensorMeta.java
package com.tencent.angel.ps.tensor;

import com.tencent.angel.common.Meta;
import com.tencent.angel.matrix.MatrixContext;
import com.tencent.angel.model.MatrixLoadContext;
import com.tencent.angel.model.MatrixSaveContext;


public class TensorMeta extends Meta {
    public TensorMeta(String name, String dtype, int dim, long[] shape, long validIndexNum) {
        super(name, dtype, dim, shape, validIndexNum);

        assert dim == shape.length;
        if (rowType.isDense()) {
            long numCols = 1L;
            for (int i = 0; i < dim; i++) {
                numCols *= shape[i];
            }

            matrixContext = new MatrixContext(name, 1, numCols, validIndexNum, -1, -1, rowType);
        } else if (rowType.isSparse()) {
            assert dim <= 2 && dim > 0;
            if (dim == 1) {
                long numCols = shape[0];
                matrixContext = new MatrixContext(name, 1, numCols, validIndexNum, -1, -1, rowType);
            } else { // dim == 2
                matrixContext = new MatrixContext(name, (int) shape[0], shape[1], validIndexNum, -1, -1, rowType);
            }
        }

    }

    public TensorMeta(String name, String dtype, int dim, long[] shape) {
        this(name, dtype, dim, shape, -1);
    }

    public TensorMeta(String name, String dtype, int numRows, long numCols) {
        this(name, dtype, numRows, numRows, -1);

    }

    public TensorMeta(String name, String dtype, int numRows, long numCols, long validIndexNum) {
        super(name, dtype, 2, new long[]{numRows, numCols}, validIndexNum);
    }

    public MatrixContext getMatrixContext() {
        return matrixContext;
    }

    public MatrixSaveContext getMatrixSaveContext(String path, String formatClassName) {
        return new MatrixSaveContext(name, formatClassName);
    }

    public MatrixLoadContext getMatrixLoadContext(String path) {
        return new MatrixLoadContext(name, path);
    }


}

File: pyps/src/main/scala/com/tencent/client/worker/ps/tensor/TensorMeta.scala
package com.tencent.client.worker.ps.tensor

import com.tencent.angel.matrix.MatrixContext
import com.tencent.angel.model.MatrixLoadContext
import com.tencent.angel.model.MatrixSaveContext
import com.tencent.client.common.Meta


class TensorMeta(name: String, dtype: String, dim: Int, shape: Array[Long], validIndexNum: Long)
  extends Meta(name, dtype, dim, shape, validIndexNum) {

  override protected val matrixContext: MatrixContext = if (rowType.isDense) {
    val numCols = shape.product

    new MatrixContext(name, 1, numCols, validIndexNum, -1, -1, rowType)
  } else if (rowType.isSparse) {
    assert(dim <= 2 && dim > 0)
    if (dim == 1) {
      val numCols = shape(0)
      new MatrixContext(name, 1, numCols, validIndexNum, -1, -1, rowType)
    } else { // dim == 2
      new MatrixContext(name, shape(0).toInt, shape(1), validIndexNum, -1, -1, rowType)
    }
  } else {
    throw new Exception("cannot create matrix!")
  }

  def this(name: String, dtype: String, dim: Int, shape: Array[Long]) {
    this(name, dtype, dim, shape, -1)
  }

  def this(name: String, dtype: String, numRows: Int, numCols: Long, validIndexNum: Long) {
    this(name, dtype, 2, Array[Long](numRows, numCols), validIndexNum)
  }

  def this(name: String, dtype: String, numRows: Int, numCols: Long) {
    this(name, dtype, numRows, numRows, -1)
  }

  def getMatrixContext: MatrixContext = {
    matrixContext
  }

  def getMatrixSaveContext(path: String, formatClassName: String): MatrixSaveContext = {
    new MatrixSaveContext(name, formatClassName)
  }

  def getMatrixLoadContext(path: String): MatrixLoadContext = {
    new MatrixLoadContext(name, path)
  }


}
